server/service/src/standard_reports.rs

```diff
@@ -41,33 +44,66 @@ impl StandardReports {
+                    ),
+                    report,
+                )
+            })
+            .collect::<HashMap<_, _>>();
 
-            if let Some(form_schema_json) = &report.form_schema {
-                // TODO: Look up existing json schema and use it's ID to be safe...
-                FormSchemaRowRepository::new(con).upsert_one(form_schema_json)?;
-            }
+        let reports_to_upsert: Vec<ReportData> = reports_data
+            .reports
+            .into_iter()
+            .filter(|report| {
+                !existing_reports.contains_key(&(
+                    report.code.clone(),
+                    report.version.clone(),
+                    report.is_custom,
+                ))
+            })
+            .collect();
+
+        for report in reports_to_upsert {
+            if !existing_reports.keys().any(|(code, version, is_custom)| {
+                code == &report.code
+                    && (Version::from(&version).unwrap() >= Version::from(&report.version).unwrap())
+                    && is_custom == &report.is_custom
+            }) {
+                num_std_reports += 1;
+                let existing_report = ReportRowRepository::new(con)
+                    .find_one_by_code_and_version(&report.code, &report.version)?;
+
+                // Use the existing ID if already defined for that report
+                let id = existing_report.map_or_else(|| report.clone().id, |r| r.id.clone());
+
+                if let Some(form_schema_json) = &report.form_schema {
+                    // TODO: Look up existing json schema and use it's ID to be safe...
+                    FormSchemaRowRepository::new(con).upsert_one(form_schema_json)?;
+                }
 
-            ReportRowRepository::new(con).upsert_one(&ReportRow {
-                id,
-                name: report.name,
-                r#type: repository::ReportType::OmSupply,
-                template: serde_json::to_string_pretty(&report.template)?,
-                context: report.context,
-                sub_context: report.sub_context,
-                argument_schema_id: report.argument_schema_id,
-                comment: report.comment,
-                is_custom: report.is_custom,
-                version: report.version,
-                code: report.code,
-            })?;
+                ReportRowRepository::new(con).upsert_one(&ReportRow {
+                    id,
+                    name: report.name,
+                    r#type: repository::ReportType::OmSupply,
+                    template: serde_json::to_string_pretty(&report.template)?,
+                    context: report.context,
+                    sub_context: report.sub_context,
+                    argument_schema_id: report.argument_schema_id,
+                    comment: report.comment,
+                    is_custom: report.is_custom,
+                    version: report.version,
+                    code: report.code,
+                })?;
+            };
```

Comment:
Hmm not very happy with logic here, do you mind doing a self review of the changes and existing logic in this file and seeing how it can/should be simplified ?

Reply:
Hi Andrei,

I was wondering about just upserting all reports regardless of if they were older versions than the one existing on the server. 
Is this what you're thinking? Or are you thinking refactoring but keeping the current logic?